File: langs/typescript/src/models/scenario.ts
import type { DecisionM } from './decision.js';
import type { AlternativeM } from './alternative.js';
import type { PropertyM } from './property.js';
import type { CoefficientM } from './coefficient.js';
import type { ConstraintM } from './constraint.js';
import type { ConfigM } from './config.js';

export interface ScenarioM {
  readonly schemaVersion: string;
  readonly name: string;
  readonly description: string;
  readonly decisions: readonly DecisionM[];
  readonly alternatives: readonly AlternativeM[];
  readonly properties: readonly PropertyM[];
  readonly coefficients: readonly CoefficientM[];
  readonly constraints: readonly ConstraintM[];
  readonly config: ConfigM;
  readonly warnings: readonly string[];
}

File: langs/python/src/guidearch/models/scenario.py
"""Scenario domain model — top-level container for a TOPSIS scenario."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from guidearch.models.alternative import AlternativeM
from guidearch.models.coefficient import CoefficientM
from guidearch.models.constraint import Constraint
from guidearch.models.decision import DecisionM
from guidearch.models.normalized_fuzzy import NormalizedFuzzyM
from guidearch.models.property import PropertyM

# Weights triple: (positive, average, negative) summing to 1.
# Reuses the NormalizedFuzzyM shape (positive/average/negative fields).
Weights = NormalizedFuzzyM


@dataclass(frozen=True)
class ConfigM:
    """Scenario configuration.

    aggregation: 'sum' or 'max' (PhiCalculationApproach); default 'max'.
    weights: (w+, wa, w-) summing to 1 (SolutionApproach presets).
    """

    aggregation: Literal["sum", "max"]
    weights: Weights


@dataclass(frozen=True)
class ScenarioM:
    """Complete decision-analysis scenario for fuzzy TOPSIS ranking."""

    schema_version: str
    name: str
    description: str
    decisions: tuple[DecisionM, ...]
    alternatives: tuple[AlternativeM, ...]
    properties: tuple[PropertyM, ...]
    coefficients: tuple[CoefficientM, ...]
    constraints: tuple[Constraint, ...]
    config: ConfigM
    warnings: tuple[str, ...] = field(default_factory=tuple)
